controllers/flinkcluster_reconciler.go

```diff
@@ -428,23 +424,50 @@ func (reconciler *ClusterReconciler) getFlinkJobID() string {
 	return ""
 }
 
-func (reconciler *ClusterReconciler) shouldTakeSavepoint() bool {
-	var jobSpec = reconciler.observed.cluster.Spec.Job
+func (reconciler *ClusterReconciler) isJobFinished() bool {
 	var jobStatus = reconciler.observed.cluster.Status.Components.Job
+	return jobStatus != nil &&
+		(jobStatus.State == v1alpha1.JobState.Succeeded ||
+			jobStatus.State == v1alpha1.JobState.Failed ||
+			jobStatus.State == v1alpha1.JobState.Cancelled)
+}
 
-	// Not enabled.
-	if jobSpec.AutoSavepointSeconds == nil || jobSpec.SavepointsDir == nil {
-		return false
+// Takes a savepoint if possible then stops the job.
+func (reconciler *ClusterReconciler) cancelJob(jobID string) error {
+	var log = reconciler.log
+	if reconciler.canTakeSavepoint() {
+		var err = reconciler.takeSavepoint(jobID)
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Info("Can not take savepoint before stopping job", "jobID", jobID)
 	}
+	if !reconciler.isJobFinished() {
+		return reconciler.stopJob(jobID)
+	}
+	log.Info("Job has finished, no need to cancel", "jobID", jobID)
+	return nil
+}
+
+// Stops a job through Flink REST API.
+func (reconciler *ClusterReconciler) stopJob(jobID string) error {
```

Comment:
Worth creating a wrapper method here? seems like it's only called once in this method and whatever pre-logic here can just go there.